### PR TITLE
rbd: add backend support for VolumeGroup operations

### PR DESCRIPTION
**Pull request by nixpanic**
# Describe what this PR does #

Implementation of the VolumeGroup backend. It contains a clear separation of the VolumeGroup type and functions and CSI-Addons responsibilities.

## Is there anything that requires special attention ##

Tested with a [small application that calls the CSI-Addons procedures directly](https://gist.github.com/nixpanic/3aab1eb6ad1937a52fb289e91ace3272). The application was modified several times to test different scenarios. At the moment I am confident that the basic functionality works as expected.

```
I0722 14:28:27.040708       1 utils.go:235] ID: 22 GRPC call: /volumegroup.Controller/CreateVolumeGroup
I0722 14:28:27.040980       1 utils.go:236] ID: 22 GRPC request: {"name":"my-group","parameters":{"clusterID":"openshift-storage","pool":"ocs-storagecluster-cephblockpool"},"secrets":"***stripped***","volume_ids":["0001-0011-openshift-storage-0000000000000001-eafd9e22-70de-459b-97fc-2d46e03d5f88","0001-0011-openshift-storage-0000000000000001-2f1acca8-95ca-4594-be0c-e5e70b6cbc43"]}
I0722 14:28:27.065729       1 omap.go:89] ID: 22 got omap values: (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.eafd9e22-70de-459b-97fc-2d46e03d5f88"): map[csi.imageid:67c16fdb4f5d csi.imagename:csi-vol-eafd9e22-70de-459b-97fc-2d46e03d5f88 csi.volname:pvc-37249c71-61f2-4c2f-8cdd-08cb2f99d668 csi.volume.owner:default]
I0722 14:28:27.112579       1 omap.go:89] ID: 22 got omap values: (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.2f1acca8-95ca-4594-be0c-e5e70b6cbc43"): map[csi.imageid:67c1ec38b881 csi.imagename:csi-vol-2f1acca8-95ca-4594-be0c-e5e70b6cbc43 csi.volname:pvc-5aae7b38-7ec3-4c45-b3c9-1a21c1d37b78 csi.volume.owner:default]
I0722 14:28:27.143738       1 volumegroup.go:106] ID: 22 all 2 Volumes for VolumeGroup "my-group" have been found
I0722 14:28:27.156146       1 omap.go:159] ID: 22 set omap keys (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.groups."): map[csi.volume.group.my-group:46ae4226-e71b-4955-8440-4801f178979a])
I0722 14:28:27.160879       1 omap.go:159] ID: 22 set omap keys (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.group.46ae4226-e71b-4955-8440-4801f178979a"): map[csi.groupname:csi-vol-group-46ae4226-e71b-4955-8440-4801f178979a csi.volname:my-group])
I0722 14:28:27.161611       1 omap.go:221] ID: 22 got omap values: (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.group.46ae4226-e71b-4955-8440-4801f178979a"): map[csi.groupname:csi-vol-group-46ae4226-e71b-4955-8440-4801f178979a csi.volname:my-group]
I0722 14:28:27.161663       1 volume_group.go:149] ID: 22 GetVolumeGroup(0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a) returns {id:0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a name:csi-vol-group-46ae4226-e71b-4955-8440-4801f178979a clusterID:openshift-storage credentials:0xc0008111a0 conn:<nil> ioctx:<nil> monitors:172.30.253.8:3300,172.30.149.93:3300,172.30.68.64:3300 pool:ocs-storagecluster-cephblockpool namespace: journal:0xc000b2e840 volumes:[] volumesToFree:[]}
I0722 14:28:27.161697       1 volume_group.go:249] ID: 22 connection established for volume group "0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a"
I0722 14:28:27.161710       1 volume_group.go:277] ID: 22 iocontext created for volume group "0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a" in pool "ocs-storagecluster-cephblockpool"
I0722 14:28:27.170848       1 volume_group.go:326] ID: 22 volume group "csi-vol-group-46ae4226-e71b-4955-8440-4801f178979a" has been created
I0722 14:28:27.170869       1 volumegroup.go:118] ID: 22 VolumeGroup "my-group" had been created: csi-vol-group-46ae4226-e71b-4955-8440-4801f178979a
I0722 14:28:27.194642       1 omap.go:159] ID: 22 set omap keys (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.group.46ae4226-e71b-4955-8440-4801f178979a"): map[0001-0011-openshift-storage-0000000000000001-eafd9e22-70de-459b-97fc-2d46e03d5f88:])
I0722 14:28:27.216181       1 omap.go:159] ID: 22 set omap keys (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.group.46ae4226-e71b-4955-8440-4801f178979a"): map[0001-0011-openshift-storage-0000000000000001-2f1acca8-95ca-4594-be0c-e5e70b6cbc43:])
I0722 14:28:27.216203       1 volumegroup.go:133] ID: 22 all 2 Volumes have been added to for VolumeGroup "my-group"
I0722 14:28:27.216519       1 utils.go:242] ID: 22 GRPC response: {"volume_group":{"volume_group_id":"0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a","volumes":[{"capacity_bytes":17179869184,"volume_context":{"imageName":"csi-vol-eafd9e22-70de-459b-97fc-2d46e03d5f88","journalPool":"ocs-storagecluster-cephblockpool","pool":"ocs-storagecluster-cephblockpool"},"volume_id":"0001-0011-openshift-storage-0000000000000001-eafd9e22-70de-459b-97fc-2d46e03d5f88"},{"capacity_bytes":8589934592,"volume_context":{"imageName":"csi-vol-2f1acca8-95ca-4594-be0c-e5e70b6cbc43","journalPool":"ocs-storagecluster-cephblockpool","pool":"ocs-storagecluster-cephblockpool"},"volume_id":"0001-0011-openshift-storage-0000000000000001-2f1acca8-95ca-4594-be0c-e5e70b6cbc43"}]}}
I0722 14:28:27.217582       1 utils.go:235] ID: 23 GRPC call: /volumegroup.Controller/DeleteVolumeGroup
I0722 14:28:27.217667       1 utils.go:236] ID: 23 GRPC request: {"secrets":"***stripped***","volume_group_id":"0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a"}
I0722 14:28:27.218503       1 omap.go:221] ID: 23 got omap values: (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.group.46ae4226-e71b-4955-8440-4801f178979a"): map[0001-0011-openshift-storage-0000000000000001-2f1acca8-95ca-4594-be0c-e5e70b6cbc43: 0001-0011-openshift-storage-0000000000000001-eafd9e22-70de-459b-97fc-2d46e03d5f88: csi.groupname:csi-vol-group-46ae4226-e71b-4955-8440-4801f178979a csi.volname:my-group]
I0722 14:28:27.219984       1 omap.go:89] ID: 23 got omap values: (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.2f1acca8-95ca-4594-be0c-e5e70b6cbc43"): map[csi.imageid:67c1ec38b881 csi.imagename:csi-vol-2f1acca8-95ca-4594-be0c-e5e70b6cbc43 csi.volname:pvc-5aae7b38-7ec3-4c45-b3c9-1a21c1d37b78 csi.volume.owner:default]
I0722 14:28:27.258472       1 omap.go:89] ID: 23 got omap values: (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.eafd9e22-70de-459b-97fc-2d46e03d5f88"): map[csi.imageid:67c16fdb4f5d csi.imagename:csi-vol-eafd9e22-70de-459b-97fc-2d46e03d5f88 csi.volname:pvc-37249c71-61f2-4c2f-8cdd-08cb2f99d668 csi.volume.owner:default]
I0722 14:28:27.289357       1 volume_group.go:149] ID: 23 GetVolumeGroup(0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a) returns {id:0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a name:csi-vol-group-46ae4226-e71b-4955-8440-4801f178979a clusterID:openshift-storage credentials:0xc000c832c0 conn:<nil> ioctx:<nil> monitors:172.30.253.8:3300,172.30.149.93:3300,172.30.68.64:3300 pool:ocs-storagecluster-cephblockpool namespace: journal:0xc00016f640 volumes:[0xc000658b48 0xc000658d88] volumesToFree:[0xc000658b48 0xc000658d88]}
I0722 14:28:27.289378       1 volumegroup.go:185] ID: 23 VolumeGroup "0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a" has been found
I0722 14:28:27.289387       1 volumegroup.go:197] ID: 23 VolumeGroup "0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a" contains 2 volumes
I0722 14:28:27.289420       1 volume_group.go:307] ID: 23 destroyed volume group instance with id "0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a"
E0722 14:28:27.289438       1 utils.go:240] ID: 23 GRPC error: rpc error: code = FailedPrecondition desc = rejecting to delete non-empty volume group "0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a"
```

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

### CODE REVIEW

File: internal/rbd/controllerserver.go
Comment (Madhu-1) [inline internal/rbd/controllerserver.go]: panic or return error here as well? continue will result in other problems

File: internal/rbd/group.go
Comment (Madhu-1) [inline internal/rbd/group.go:25]: rearrange the order here? cephcsi import should come first

File: internal/rbd/group.go
Comment (Madhu-1) [inline internal/rbd/group.go]: `image to the group` to `image from the group`

File: internal/rbd/group.go
Comment (Madhu-1) [inline internal/rbd/group.go:68]: ioctx is not used

File: internal/rbd/group.go
Comment (Madhu-1) [inline internal/rbd/group.go:31]: ioctx is not used

File: internal/rbd/manager.go
Comment (Madhu-1) [inline internal/rbd/manager.go:143]: i think in case of any error we need to destroy the connection to journal here

File: internal/rbd/manager.go
Comment (Madhu-1) [inline internal/rbd/manager.go:172]: same here as well Destory is required?

File: internal/rbd/manager.go
Comment (Madhu-1) [inline internal/rbd/manager.go:177]: validate `pool` is not empty string?

File: internal/rbd_group/volume_group.go
Comment (Madhu-1) [inline internal/rbd_group/volume_group.go]: reorder is required here

File: internal/rbd/group.go
Comment (nixpanic) [reply internal/rbd/group.go:25]: this guideline is not followed everywere. I think and Golang suggests to have imports of local packages at the bottom, that seems to be a coding convention many other Go projects use.

File: internal/rbd/controllerserver.go
Comment (nixpanic) [reply internal/rbd/controllerserver.go]: Do you have a preference to panic or return an error? I guess returning an error is the cleanest.

File: internal/rbd/group.go
Comment (nixpanic) [reply internal/rbd/group.go:31]: yes it is, on `return librbd.GroupImageAdd(ioctx, ...`.

File: internal/rbd/group.go
Comment (nixpanic) [reply internal/rbd/group.go:68]: yes it is, on `return librbd.GroupImageAdd(ioctx, ...`.

File: internal/rbd/manager.go
Comment (nixpanic) [reply internal/rbd/manager.go:143]: The Manager takes care of that during `Destroy()`. This makes the code simpler as there are no `defer` statements needed that check for errors.

File: internal/rbd/manager.go
Comment (nixpanic) [reply internal/rbd/manager.go:172]: no, see above.

File: internal/rbd_group/volume_group.go
Comment (nixpanic) [reply internal/rbd_group/volume_group.go]: I don't like that :wink: , many Go projects prefer this ordering. It would be nice if Ceph-CSI can use a similar standard formatting.

File: internal/rbd/controllerserver.go
Comment (Madhu-1) [reply internal/rbd/controllerserver.go]: return error looks cleaner

File: internal/rbd/group.go
Comment (Madhu-1) [reply internal/rbd/group.go:25]: https://github.com/ceph/ceph-csi/blob/devel/docs/coding.md#imports this is one we have, lets stick to it and also update other places if its not followed as followup cleanup?

File: internal/rbd/group.go
Comment (Madhu-1) [reply internal/rbd/group.go:31]: but you are passing rv.ioctx not the one you get from here

Comment (Madhu-1) [reply internal/rbd/group.go:31]: ha my mistake, it looks like we have two ioctx

File: internal/rbd/controllerserver.go
Comment (nixpanic) [reply internal/rbd/controllerserver.go]: done

File: internal/rbd/group.go
Comment (nixpanic) [reply internal/rbd/group.go:25]: See #4721. I don't think we should strictly enforce it for existing imports, but it would be nice to use this order in new files.

**Review by Madhu-1 — COMMENTED** [review]
LGTM, let me know once its tested, i will do final review and approve it.

**Comment by yati1998** [comment]
LGTM

File: internal/rbd/manager.go
Comment (Rakshith-R) [inline internal/rbd/manager.go]: https://github.com/ceph/ceph-csi/blob/devel/internal/journal/volumegroupjournal.go#L170 

checkreservation needs to be called before trying to reserve the name ? 

similar to https://github.com/ceph/ceph-csi/blob/devel/internal/cephfs/store/volumegroup.go#L199

Comment (nixpanic) [reply internal/rbd/manager.go]: this is done now, it seems to be working well with the added `CheckReservation` call.

File: internal/csi-addons/rbd/volumegroup.go
Comment (Rakshith-R) [inline internal/csi-addons/rbd/volumegroup.go]:
```
	log.DebugLog(ctx, fmt.Sprintf("VolumeGroup %q has been created: %+v", req.GetName(), vg))
Comment
```

**Review by Rakshith-R — COMMENTED** [review]
Tiny nits.

There's 3 commits named `rbd: implement the VolumeGroup interface` that I think can be squashed together.

File: internal/csi-addons/rbd/volumegroup.go
Comment (nixpanic) [reply internal/csi-addons/rbd/volumegroup.go]: @Rakshith-R, sorry, I don't understand what you mean by this?

The `fmt.Sprintf()` call can be dropped, is that it?

**Review by Rakshith-R — APPROVED** [review]
Thanks LGTM

**Review by yati1998 — APPROVED** [review]

**Comment by nixpanic** [comment]
@mergifyio rebase

**Comment by nixpanic** [comment]
@mergifyio queue

**Comment by nixpanic** [comment]
@mergifyio requeue

The **multi-arch-build** CI job failed while installing some packages. It has been restarted and has passed the section that previously failed.

**Pull request merged**
